File: Frontend/src/lib/api/production-orders.ts
import { apiClient } from './client';

// Types
export interface ProductionOrder {
  id: string;
  order_number: string;
  product_id: string;
  product_code: string;
  product_name: string;
  quantity: number;
  unit: string;
  due_date: string;
  priority: string;
  status: string;
  notes?: string;
  qr_code?: string;
  materials_status?: string;
  days_until_due?: number;
  is_overdue?: boolean;
  materials?: OrderMaterial[];
  total_material_cost?: number;
  started_at?: string;
  completed_at?: string;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface OrderMaterial {
  id: string;
  order_id: string;
  material_id: string;
  material_code?: string;
  material_name?: string;
  required_qty: number;
  allocated_qty: number;
  issued_qty: number;
  unit: string;
  unit_cost: number;
  total_cost: number;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface CreateProductionOrderData {
  product_id: string;
  quantity: number;
  due_date: string;
  priority: string;
  notes?: string;
  customer_name?: string;
  assigned_team?: string;
  shift_number?: string;
  production_stage?: string;
  start_time?: string;
  end_time?: string;
}

export interface UpdateProductionOrderData {
  quantity?: number;
  due_date?: string;
  priority?: string;
  status?: string;
  notes?: string;
  customer_name?: string;
  assigned_team?: string;
  shift_number?: string;
  production_stage?: string;
}

export interface OrderStatusUpdate {
  status: string;
  notes?: string;
}

export interface MaterialRequirement {
  material_id: string;
  material_code: string;
  material_name: string;
  required_qty: number;
  available_qty: number;
  shortage_qty: number;
  unit: string;
  availability_status: string;
}

export interface OrderProgress {
  order_id: string;
  order_number: string;
  current_stage: string;
  overall_progress: number;
  stages: StageProgress[];
}

export interface StageProgress {
  stage_name: string;
  status: string;
  progress: number;
  started_at?: string;
  completed_at?: string;
  assigned_team?: string;
}

export interface TeamAssignment {
  user_id: string;
  user_name: string;
  role: string;
  assigned_at: string;
}

export interface ProductionValidation {
  can_produce: boolean;
  product_id: string;
  product_name: string;
  requested_quantity: number;
  shortages: MaterialRequirement[];
  total_shortage_value: number;
  message: string;
}

export interface ListOrdersParams {
  page?: number;
  limit?: number;
  status?: string;
  priority?: string;
  search?: string;
  due_date_from?: string;
  due_date_to?: string;
}

// API Client
export const productionOrdersApi = {
  // List production orders
  async listOrders(params?: ListOrdersParams): Promise<ProductionOrder[]> {
    const queryParams = new URLSearchParams();
    
    if (params?.page) queryParams.append('page', params.page.toString());
    if (params?.limit) queryParams.append('limit', params.limit.toString());
    if (params?.status) queryParams.append('status', params.status);
    if (params?.priority) queryParams.append('priority', params.priority);
    if (params?.search) queryParams.append('search', params.search);
    if (params?.due_date_from) queryParams.append('due_date_from', params.due_date_from);
    if (params?.due_date_to) queryParams.append('due_date_to', params.due_date_to);

    const url = `/production-orders${queryParams.toString() ? '?' + queryParams.toString() : ''}`;
    return await apiClient.get<ProductionOrder[]>(url);
  },

  // Get single production order
  async getOrder(orderId: string): Promise<ProductionOrder> {
    return await apiClient.get<ProductionOrder>(`/production-orders/${orderId}`);
  },

  // Create production order
  async createOrder(data: CreateProductionOrderData): Promise<ProductionOrder> {
    return await apiClient.post<ProductionOrder>('/production-orders', data);
  },

  // Duplicate production order
  async duplicateOrder(orderId: string): Promise<ProductionOrder> {
    return await apiClient.post<ProductionOrder>(`/production-orders/${orderId}/duplicate`);
  },

  // Update production order
  async updateOrder(orderId: string, data: UpdateProductionOrderData): Promise<ProductionOrder> {
    return await apiClient.put<ProductionOrder>(`/production-orders/${orderId}`, data);
  },

  // Update order status
  async updateStatus(orderId: string, data: OrderStatusUpdate): Promise<ProductionOrder> {
    return await apiClient.put<ProductionOrder>(`/production-orders/${orderId}/status`, data);
  },

  // Archive order
  async archiveOrder(orderId: string): Promise<ProductionOrder> {
    return await apiClient.post<ProductionOrder>(`/production-orders/${orderId}/archive`);
  },

  // Cancel/Delete order
  async cancelOrder(orderId: string): Promise<{ message: string }> {
    return await apiClient.delete<{ message: string }>(`/production-orders/${orderId}`);
  },

  // Get order materials
  async getOrderMaterials(orderId: string): Promise<MaterialRequirement[]> {
    return await apiClient.get<MaterialRequirement[]>(`/production-orders/${orderId}/materials`);
  },

  // Get order progress
  async getOrderProgress(orderId: string): Promise<OrderProgress> {
    return await apiClient.get<OrderProgress>(`/production-orders/${orderId}/progress`);
  },

  // Assign team to order
  async assignTeam(orderId: string, userIds: string[]): Promise<{ message: string }> {
    return await apiClient.post<{ message: string }>(`/production-orders/${orderId}/assign`, { user_ids: userIds });
  },

  // Get order team
  async getOrderTeam(orderId: string): Promise<TeamAssignment[]> {
    return await apiClient.get<TeamAssignment[]>(`/production-orders/${orderId}/team`);
  },

  // Validate production feasibility
  async validateProduction(productId: string, quantity: number, locationId?: string): Promise<ProductionValidation> {
    const queryParams = new URLSearchParams({
      product_id: productId,
      quantity: quantity.toString(),
    });
    
    if (locationId) {
      queryParams.append('target_location_id', locationId);
    }

    return await apiClient.post<ProductionValidation>(`/production-orders/validate-production?${queryParams.toString()}`);
  },
};

File: Backend/app/api/v1/endpoints/production_orders.py
from fastapi import APIRouter, Depends, Query
from typing import List, Optional
from app.schemas.production_order import (
    ProductionOrderCreate, ProductionOrderUpdate, ProductionOrderResponse,
    ProductionOrderListItem, MaterialRequirement, OrderProgress, OrderStatusUpdate,
    TeamAssignment, ProductionOrderValidation
)
from app.schemas.user import UserResponse
from app.services.production_order_service import production_order_service
from app.api.deps import get_current_user, require_role
from decimal import Decimal

router = APIRouter()


@router.post("/", response_model=ProductionOrderResponse, status_code=201)
async def create_production_order(
    order_data: ProductionOrderCreate,
    current_user: UserResponse = Depends(require_role("Planner"))
):
    """
    Create a new production order.
    
    - Auto-generates order number (PO-YYYY-XXXX)
    - Calculates material requirements from BOM
    - Checks material availability
    - **product_id**: Finished goods product
    - **quantity**: Order quantity
    - **due_date**: Target completion date
    - **priority**: Low/Medium/High/Urgent
    - **customer_name**: Customer name (optional)
    - **production_stage**: Current stage (optional)
    - **shift_number**: Shift assignment (optional)
    - **start_time**: Production start time (optional)
    - **end_time**: Production end time (optional)
    """
    return await production_order_service.create_production_order(order_data, current_user.id)


@router.post("/{order_id}/duplicate", response_model=ProductionOrderResponse, status_code=201)
async def duplicate_production_order(
    order_id: str,
    current_user: UserResponse = Depends(require_role("Planner"))
):
    """
    Duplicate an existing production order.
    
    Creates a new order with:
    - Same product and quantity
    - Same BOM and material requirements
    - New order number
    - Due date set to 7 days from now
    - Status: Planned
    """
    return await production_order_service.duplicate_production_order(order_id, current_user.id)


@router.get("/", response_model=List[ProductionOrderListItem])
async def list_production_orders(
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=100),
    status: Optional[str] = None,
    priority: Optional[str] = None,
    search: Optional[str] = None,
    due_date_from: Optional[str] = None,
    due_date_to: Optional[str] = None,
    current_user: UserResponse = Depends(get_current_user)
):
    """
    List all production orders with filters.
    
    - **status**: Planned/In Progress/Completed/Cancelled
    - **priority**: Low/Medium/High/Urgent
    - **search**: Search in order number or product name
    - **due_date_from/to**: Date range filter
    """
    return await production_order_service.list_production_orders(
        page, limit, status, priority, search, due_date_from, due_date_to
    )


@router.get("/{order_id}", response_model=ProductionOrderResponse)
async def get_production_order(
    order_id: str,
    current_user: UserResponse = Depends(get_current_user)
):
    """
    Get production order details with material requirements.
    """
    return await production_order_service.get_order_by_id(order_id)


@router.put("/{order_id}", response_model=ProductionOrderResponse)
async def update_production_order(
    order_id: str,
    order_data: ProductionOrderUpdate,
    current_user: UserResponse = Depends(require_role("Planner"))
):
    """
    Update production order (only if status is Planned).
    """
    return await production_order_service.update_production_order(order_id, order_data, current_user.id)


@router.put("/{order_id}/status", response_model=ProductionOrderResponse)
async def update_order_status(
    order_id: str,
    status_data: OrderStatusUpdate,
    current_user: UserResponse = Depends(require_role("Supervisor"))
):
    """
    Update order status.
    
    Workflow: Planned → In Progress → Completed
    Can cancel from any state
    """
    return await production_order_service.update_order_status(order_id, status_data, current_user.id)


@router.post("/{order_id}/archive", response_model=ProductionOrderResponse)
async def archive_production_order(
    order_id: str,
    current_user: UserResponse = Depends(require_role("Planner"))
):
    """
    Archive a production order.
    
    Only completed or cancelled orders can be archived.
    Archived orders are hidden from main list but can be retrieved.
    """
    return await production_order_service.archive_production_order(order_id, current_user.id)


@router.delete("/{order_id}")
async def cancel_production_order(
    order_id: str,
    current_user: UserResponse = Depends(require_role("Planner"))
):
    """
    Cancel production order.
    
    Cannot cancel if materials already issued or WIP exists.
    """
    return await production_order_service.cancel_production_order(order_id, current_user.id)


@router.get("/{order_id}/materials", response_model=List[MaterialRequirement])
async def get_order_materials(
    order_id: str,
    current_user: UserResponse = Depends(get_current_user)
):
    """
    Get calculated material requirements for order.
    
    Shows:
    - Required quantity (with scrap %)
    - Available quantity
    - Shortage quantity
    - Availability status
    """
    return await production_order_service.get_order_materials(order_id)


@router.get("/{order_id}/progress", response_model=OrderProgress)
async def get_order_progress(
    order_id: str,
    current_user: UserResponse = Depends(get_current_user)
):
    """
    Get order progress through stages.
    
    Returns stage completion status and timelines.
    """
    return await production_order_service.get_order_progress(order_id)


@router.post("/{order_id}/assign")
async def assign_team_to_order(
    order_id: str,
    user_ids: List[str],
    current_user: UserResponse = Depends(require_role("Supervisor"))
):
    """
    Assign team members to production order.
    """
    return await production_order_service.assign_team(order_id, user_ids, current_user.id)


@router.get("/{order_id}/team", response_model=List[TeamAssignment])
async def get_order_team(
    order_id: str,
    current_user: UserResponse = Depends(get_current_user)
):
    """
    Get team assignments for order.
    """
    return await production_order_service.get_team_assignments(order_id)

@router.post("/validate-production", response_model=ProductionOrderValidation)
async def validate_production_feasibility(
    product_id: str = Query(..., description="Finished goods product ID"),
    quantity: Decimal = Query(..., gt=0, description="Production quantity"),
    target_location_id: Optional[str] = Query(None, description="Check inventory at specific location"),
    current_user: UserResponse = Depends(get_current_user)
):
    """
    Validate if production order can be fulfilled with current inventory.
    
    **Features:**
    - Checks material availability before creating order
    - Shows exact shortage quantities
    - Returns feasibility status (can_produce: true/false)
    - Helps with production planning
    
    **Use Cases:**
    - Pre-production validation
    - Material planning
    - Order feasibility check
    - Quick shortage preview
    
    **Parameters:**
    - **product_id**: Finished goods product ID
    - **quantity**: Production quantity to validate
    - **target_location_id**: (Optional) Check specific warehouse
    
    **Response:**
    - can_produce: true/false
    - Material-wise shortage details
    - Summary of procurement needs
    """
    return await production_order_service.validate_production_feasibility(
        product_id=product_id,
        quantity=quantity,
        target_location_id=target_location_id
    )